cmd/registry/cmd/label.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/apigee/registry/cmd/registry/core"
+	"github.com/apigee/registry/connection"
+	"github.com/apigee/registry/gapic"
+	"github.com/apigee/registry/rpc"
+	"github.com/apigee/registry/server/names"
+	"github.com/spf13/cobra"
+	"google.golang.org/genproto/protobuf/field_mask"
+)
+
+var labelFilter string
+var labelPairs map[string]string
+
+func init() {
+	rootCmd.AddCommand(labelCmd)
+	labelCmd.Flags().StringVar(&labelFilter, "filter", "", "Filter selected resources")
+}
+
+var labelCmd = &cobra.Command{
+	Use:   "label RESOURCE KEY_1=VAL_1 ... KEY_N=VAL_N",
+	Short: "Label resources in the API Registry",
+	Args:  cobra.MinimumNArgs(2),
+	Run: func(cmd *cobra.Command, args []string) {
+		ctx := context.TODO()
+		client, err := connection.NewClient(ctx)
+		if err != nil {
+			log.Fatalf("%s", err.Error())
+		}
+
+		// Initialize task queue.
+		taskQueue := make(chan core.Task, 1024)
+		workerCount := 64
+		for i := 0; i < workerCount; i++ {
+			core.WaitGroup().Add(1)
+			go core.Worker(ctx, taskQueue)
+		}
+
+		if len(args) > 1 {
+			labelPairs = make(map[string]string)
+			for _, labeling := range args[1:] {
+				pair := strings.Split(labeling, "=")
+				if len(pair) != 2 {
+					log.Fatalf("%q must have the form \"key=value\" (value can be empty)", labeling)
+				}
+				if pair[0] == "" {
+					log.Fatalf("%q is invalid because it specifies an empty label", labeling)
+				}
+				labelPairs[pair[0]] = pair[1]
+			}
+		}
+
+		err = matchAndHandleLabelCmd(ctx, client, taskQueue, args[0])
+		if err != nil {
+			log.Fatalf("%s", err.Error())
+		}
+
+		close(taskQueue)
+		core.WaitGroup().Wait()
+	},
+}
+
+type labelApiTask struct {
+	ctx    context.Context
+	client connection.Client
+	api    *rpc.Api
+}
+
+type labelVersionTask struct {
+	ctx     context.Context
+	client  connection.Client
+	version *rpc.ApiVersion
+}
+
+type labelSpecTask struct {
+	ctx    context.Context
+	client connection.Client
+	spec   *rpc.ApiSpec
+}
+
+func (task *labelApiTask) Name() string {
+	return "label " + task.api.Name
+}
+
+func (task *labelApiTask) Run() error {
+	if task.api.Labels == nil {
+		task.api.Labels = make(map[string]string)
+	}
+	if len(labelPairs) > 0 {
+		log.Printf("label %s with %v", task.api.Name, labelPairs)
+		for k, v := range labelPairs {
+			if v == "" {
```

Comment:
Can you explain your reasoning for this behavior? If I'm understanding correctly, it means this command will ignore any keys set to empty values such as `someKey=`.

I think our API has different behavior by allowing empty label values, so this seems odd to me.

Reply:
This would clear labels, but I'm looking more closely at kubectl patterns (which are different) and will update this to match those.

Reply:
Okay. Our API behavior currently considers the labels field an atomic map that gets replaced fully if modified, so deleting is equivalent to not providing the entry. I'm interested in hearing if the kubectl patterns differ in any way.

Reply:
Right (about the API behavior). kubectl takes care of that when users add labels (with "key=value") or remove them (with "key-") - that's the pattern that I'm following along with also requiring "--overwrite" to modify existing labels (as a safety precaution).